src/mapping/query/function/parse_duration.rs

```diff
@@ -0,0 +1,176 @@
+use super::prelude::*;
+use lazy_static::lazy_static;
+use regex::Regex;
+use rust_decimal::{prelude::ToPrimitive, Decimal};
+use std::collections::HashMap;
+use std::str::FromStr;
+
+lazy_static! {
+    static ref RE: Regex =
+        Regex::new(r"(?i)\A(?P<value>[0-9]*\.?[0-9]+)\s?(?P<unit>[a-z]{1,2})\z").unwrap();
+    static ref UNITS: HashMap<String, Decimal> = vec![
+        ("ns", Decimal::new(1, 9)),
+        ("µs", Decimal::new(1, 6)),
+        ("ms", Decimal::new(1, 3)),
+        ("cs", Decimal::new(1, 2)),
+        ("ds", Decimal::new(1, 1)),
+        ("s", Decimal::new(1, 0)),
+        ("m", Decimal::new(60, 0)),
+        ("h", Decimal::new(3_600, 0)),
+        ("d", Decimal::new(86_400, 0)),
+    ]
+    .into_iter()
+    .map(|(k, v)| (k.to_owned(), v))
+    .collect();
+}
+
+#[derive(Debug)]
+pub(in crate::mapping) struct ParseDurationFn {
+    query: Box<dyn Function>,
+    format: Box<dyn Function>,
+}
+
+impl ParseDurationFn {
+    #[cfg(test)]
+    pub(in crate::mapping) fn new(query: Box<dyn Function>, format: &str) -> Self {
+        let format = Box::new(Literal::from(Value::from(format)));
+
+        Self { query, format }
+    }
+}
+
+impl Function for ParseDurationFn {
+    fn execute(&self, ctx: &Event) -> Result<Value> {
+        let value = {
+            let bytes = required!(ctx, self.query, Value::Bytes(v) => v);
+            String::from_utf8_lossy(&bytes).into_owned()
+        };
+
+        let fmt_decimal = {
```

Comment:
As in another comment, this is a conversion factor, not a format as such.

Reply:
Agreed, I had some trouble finding the right terminology. Will update.